src/kvstore/raftex/RaftPart.cpp

```diff
@@ -1548,56 +1548,92 @@ void RaftPart::processAppendLogRequest(
     }
 
     // req.get_last_log_id_sent() >= committedLogId_
-    if (lastLogTerm_ > 0 && req.get_last_log_term_sent() != lastLogTerm_) {
-        LOG(INFO) << idStr_ << "The local last log term is " << lastLogTerm_
-                << ", which is different from the leader's prevLogTerm "
-                << req.get_last_log_term_sent()
-                << ", the prevLogId is " << req.get_last_log_id_sent()
-                << ". So need to rollback to last committedLogId_ " << committedLogId_;
-        if (wal_->rollbackToLog(committedLogId_)) {
+    if (req.get_last_log_id_sent() == lastLogId_ && req.get_last_log_term_sent() == lastLogTerm_) {
+        // nothing to do
+        // just append log later
+    } else if (req.get_last_log_id_sent() > lastLogId_) {
+        // There is a gap
+        LOG(INFO) << idStr_ << "Local is missing logs from id " << lastLogId_
+                  << ". Need to catch up";
+        resp.set_error_code(cpp2::ErrorCode::E_LOG_GAP);
+        return;
+    } else  {
+        // check the term
+        int reqLastLogTerm  = wal_->getLogTerm(req.get_last_log_id_sent());
+        if (req.get_last_log_term_sent() != reqLastLogTerm) {
+            LOG(INFO) << idStr_ << "The local log term is " << reqLastLogTerm
+                      << ", which is different from the leader's prevLogTerm "
+                      << req.get_last_log_term_sent()
+                      << ", the prevLogId is " << req.get_last_log_id_sent()
+                      << ". So ask leader to send logs from committedLogId " << committedLogId_;
+            TermID committedLogTerm = wal_->getLogTerm(committedLogId_);
+            if (committedLogTerm > 0 ) {
+                resp.set_last_log_id(committedLogId_);
+                resp.set_last_log_term(committedLogTerm);
+            }
+            resp.set_error_code(cpp2::ErrorCode::E_LOG_GAP);
+            return;
+        }
+    }
+
+    // request get_last_log_term_sent  == wal[get_last_log_id_sent].log_term
+    size_t numLogs = req.get_log_str_list().size();
+    LogID firstId = req.get_last_log_id_sent() + 1;
+
+    std::vector<cpp2::LogEntry> logEntries (req.get_log_str_list().begin(),
+                                            req.get_log_str_list().end());
+
+    // may be need to rollback wal_
+    if (!( req.get_last_log_id_sent() == wal_->lastLogId() && req.get_last_log_term_sent() == wal_->lastLogTerm())) {
```

Comment:
yes, using

>  req.get_last_log_id_sent() == lastLogId_ && req.get_last_log_term_sent() == lastLogTerm_

 to check is better, I will modify it later.

I agree that when it is true,  it not need to check term in each log.  that is why I add this condition check. 

>     if (!( req.get_last_log_id_sent() == wal_->lastLogId() && req.get_last_log_term_sent() == wal_->lastLogTerm()))
In most case,  above check  is false, therefore we don not need check term of log one by one,  just append the new log。

Reply:
> In most case, above check is false, therefore we don not need check term of log one by one, just append the new log。

Agree.

I think `lastLogId_` and `lastLogTerm_` will be better. A extreme example is, this host converts from follower -> leader -> follower again, in the leader phase, it could write some log in wal (although the case is rare).

Reply:
i get it